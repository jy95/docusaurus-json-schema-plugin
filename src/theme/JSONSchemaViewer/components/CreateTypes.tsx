import React from "react"

import TabItem from "@theme-original/TabItem"
import Tabs from "@theme-original/Tabs"

import {
  TypeLabelSwitch,
  RenderProvidedType,
  CreateValidOrInvalid,
} from "@theme/JSONSchemaViewer/components"

import {
  isSchemaComposition,
  detectedTypes,
} from "@theme/JSONSchemaViewer/utils"

import type { JSONSchema, TypeValues } from "@theme/JSONSchemaViewer/types"

// Render a single type
type SingleTypeProps = {
  schema: Exclude<JSONSchema, true | false>
  nullable?: boolean
  type: TypeValues
}

function RenderSingleType(props: SingleTypeProps): JSX.Element {
  const { schema, type, nullable } = props

  return <RenderProvidedType schema={schema} type={type} nullable={nullable} />
}

// Render multiple type
type MultipleTypesProps = {
  schema: Exclude<JSONSchema, true | false>
  nullable?: boolean
  types: {
    value: TypeValues
    label: JSX.Element
  }[]
}

function RenderMultipleTypes(props: MultipleTypesProps): JSX.Element {
  const { schema, types, nullable } = props

  return (
    <Tabs defaultValue={types[0].value} values={types} key={"multiple_types"}>
      {types.map((val) => (
        <TabItem key={val.value} value={val.value}>
          {
            <RenderProvidedType
              type={val.value}
              schema={schema}
              nullable={nullable}
            />
          }
        </TabItem>
      ))}
    </Tabs>
  )
}

type Props = {
  [x: string]: any
  schema: Exclude<JSONSchema, true | false>
}

// Entry point
export default function CreateTypes(props: Props): JSX.Element {
  const { schema } = props

  // Several possibilities
  // 1. User only used a single type (most common case)
  // 2. User used multiple types (e.g. ["string", "null"] or ["string", "number"])
  // 3. User didn't use any type and so we must "guess" what (s)he have in mind

  // Let's our function do the magic to find which types are declared or detected
  const foundTypes = detectedTypes(schema)

  // Let's cover simple cases first
  // Either a single type that could be null
  const hasNull = foundTypes.includes("null")
  if (foundTypes.length === 1 || (hasNull && foundTypes.length === 2)) {
    // Either we got the not null type (likely what the final user wants to express)
    // Either we consider first entry as fallback if it was a standalone "null"
    const firstType = foundTypes.find((s) => s !== "null") || foundTypes[0]

    return (
      <RenderSingleType schema={schema} type={firstType} nullable={hasNull} />
    )
  }

  // Second, we have multiple type provided by user
  if (foundTypes.length > 1) {
    // remove null from resultset & prepare values & labels
    const values = foundTypes
      .filter((s) => s !== "null")
      .map((type) => ({
        value: type,
        label: <TypeLabelSwitch type={type} />,
      }))

    return (
      <RenderMultipleTypes schema={schema} types={values} nullable={hasNull} />
    )
  }

  // If at the end, we cannot find a type, it likely means user put something like :
  // { "allOf": ... } or { "if": ... }

  // If we don't encounter the SchemaComposition (allOf, ...) case, let's assume it is any
  if (!isSchemaComposition(schema)) {
    return <CreateValidOrInvalid schema={schema} />
  }

  // Otherwise, we have a SchemaComposition, which will be handled by CreateNodes
  return <></>
}
