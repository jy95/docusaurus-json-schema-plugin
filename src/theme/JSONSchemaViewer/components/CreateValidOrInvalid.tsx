import React from "react"

import {
  CreateAlwaysInvalid,
  CreateAlwaysValid,
} from "@theme/JSONSchemaViewer/JSONSchemaElements"

import type { JSONSchema } from "@theme/JSONSchemaViewer/types"

type Props = {
  [x: string]: any
  schema: JSONSchema
}

// Handle the "true" / "false" cases we can find in schemas
export default function CreateValidOrInvalid(props: Props): JSX.Element {
  const { schema } = props

  /**
   * Cases to cover
   * CreateAlwaysValid = true / {}
   * CreateAlwaysInvalid = false
   */
  return (
    <>
      {schema && <CreateAlwaysValid schema={schema} />}
      {!schema && <CreateAlwaysInvalid />}
    </>
  )
}
