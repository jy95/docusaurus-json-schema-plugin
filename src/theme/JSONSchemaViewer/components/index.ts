// To avoid circular dependancies issue, let's have a single place for exports and imports

export { default as Collapsible } from "./Collapsible"

export { default as CreateNodes } from "./CreateNodes"

export { default as CreateEdge } from "./CreateEdge"

export { default as SchemaItem } from "./SchemaItem/SchemaItem"
