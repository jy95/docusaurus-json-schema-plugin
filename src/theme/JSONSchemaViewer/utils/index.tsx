// To keep my mind happy, let's have a place where I can import stuff without knowing their final place in internal structure

// Re export generateFriendlyName function
export { default as GenerateFriendlyName } from "@theme/JSONSchemaViewer/utils/generateFriendlyName"

// Re export QualifierMessages function
export { default as QualifierMessages } from "@theme/JSONSchemaViewer/utils/getQualifierMessages"

// Re export detectTypes functions
export * from "@theme/JSONSchemaViewer/utils/detectTypes"
