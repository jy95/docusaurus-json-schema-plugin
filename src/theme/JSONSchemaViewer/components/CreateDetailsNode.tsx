import React from "react"
import Translate from "@docusaurus/Translate"

import { Collapsible, CreateNodes, SchemaItem } from "./index"
import { getQualifierMessages } from "../utils"

import type { JSONSchema7 } from "json-schema"

type Props = {
  name: string
  schemaName: string
  schema: JSONSchema7
  required: boolean | string[]
  [x: string]: any
}

function createDetailsNode(props: Props): JSX.Element {
  const { name, required, schema, schemaName } = props

  const isRequired = Array.isArray(required)
    ? required.includes(name)
    : required

  let qualifierMessages = getQualifierMessages(schema)

  return (
    <SchemaItem collapsible={true} schema={schema} name={name}>
      <Collapsible
        summary={
          <>
            <strong>{name}</strong>
            <span style={{ opacity: "0.6" }}>{schemaName}</span>
            {isRequired && (
              <strong
                style={{
                  fontSize: "var(--ifm-code-font-size)",
                  color: "var(--ifm-color-danger)",
                }}
              >
                <Translate
                  values={{
                    id: "json-schema.keywords.required",
                  }}
                >
                  {"required"}
                </Translate>
              </strong>
            )}
          </>
        }
        children={
          <div style={{ marginLeft: "1rem" }}>
            {qualifierMessages !== undefined && (
              <div style={{ marginTop: ".5rem", marginBottom: ".5rem" }}>
                {qualifierMessages}
              </div>
            )}
            {schema?.description !== undefined && (
              <div style={{ marginTop: ".5rem", marginBottom: ".5rem" }}>
                {schema?.description}
              </div>
            )}
            <CreateNodes schema={schema} />
          </div>
        }
      />
    </SchemaItem>
  )
}

export default createDetailsNode
