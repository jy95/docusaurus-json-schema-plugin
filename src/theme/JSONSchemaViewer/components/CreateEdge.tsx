import React, { ReactNode } from "react"

import { SchemaItem } from "./index"
import { generateFriendlyName } from "../utils/index"

import type { JSONSchema } from "../types"

// Creates the edge or "leave" of a schema tree. Edge can branch into sub-nodes.
type EdgeProps = {
  // Name of the attribute
  name: ReactNode
  // the schema of this attribute
  schema: JSONSchema
  // Is property is required or not
  required: boolean
}

function createEdge({ name, schema, required }: EdgeProps): JSX.Element {
  const schemaName = generateFriendlyName(schema)

  if (typeof schema === "boolean") {
    return <></>
  }

  return (
    <SchemaItem
      schemaName={schemaName}
      name={name}
      schema={schema}
      required={required}
    />
  )
}

export default createEdge
