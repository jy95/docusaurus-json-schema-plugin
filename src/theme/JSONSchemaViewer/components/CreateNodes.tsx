import React from "react"

import {
  CreateObject,
  SchemaComposition,
  CreateArray,
  CreatePrimitive,
} from "../JSONSchemaElements/index"

import { isArrayType, isObjectType, isSchemaComposition } from "../utils/index"

import type { JSONSchema } from "../types"

type Props = {
  [x: string]: any
  schema: JSONSchema
}

function createNodes(props: Props): JSX.Element {
  const { schema } = props

  // In boolean case, we can do anything
  if (typeof schema === "boolean") {
    return <></>
  }

  // handle anyOf / allOf / oneOf / not
  if (isSchemaComposition(schema)) {
    return <SchemaComposition schema={schema} />
  }

  // handle array case
  if (isArrayType(schema)) {
    return <CreateArray schema={schema} />
  }

  // handle object case
  if (isObjectType(schema)) {
    return <CreateObject schema={schema} />
  }

  // Well from now; two situations
  // 1. Either it is a primitive type left (string / integer / numeric /boolean / null)
  // 2. Schema is probably invalid at this point and component won't do magic tricks
  return <CreatePrimitive schema={schema} />
}

export default createNodes
