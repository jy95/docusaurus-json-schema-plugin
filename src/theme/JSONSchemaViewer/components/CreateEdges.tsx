import React from "react"

import { SchemaItem } from "../components/index"
import { generateFriendlyName } from "../utils/index"

import type { JSONSchema7Definition } from "json-schema"

// Creates the edges or "leaves" of a schema tree. Edges can branch into sub-nodes with createDetails().
type EdgeProps = {
  name: string
  schema: JSONSchema7Definition
}

function createEdges({ name, schema }: EdgeProps): JSX.Element {
  const schemaName = generateFriendlyName(schema)

  if (typeof schema === "boolean") {
    return <></>
  }

  return (
    <SchemaItem
      collapsible={false}
      schemaName={schemaName}
      name={name}
      schema={schema}
    />
  )
}

export default createEdges
