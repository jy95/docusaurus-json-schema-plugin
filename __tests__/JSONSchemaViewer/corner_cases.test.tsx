import React from "react"
import { expect, test, describe } from "@jest/globals"
import { render, act } from "@testing-library/react"

import type { RenderResult } from "@testing-library/react"
import type { JSX } from "react"

// Type to prevent creating invalid mocks
import type {
  JSONSchema,
  JSONSchemaNS,
} from "../../src/theme/JSONSchemaViewer/types"

// Components
import { CreateNodes } from "../../src/theme/JSONSchemaViewer/components/index"
import {
  CreateArray,
  CreateObject,
} from "../../src/theme/JSONSchemaViewer/JSONSchemaElements/index"
import { detectedTypes } from "../../src/theme/JSONSchemaViewer/utils/index"

const testcases: [
  string,
  (props: { schema: any; [x: string]: any }) => JSX.Element,
  JSONSchema,
][] = [
  ["CreateNodes - boolean schema (false)", CreateNodes, false],
  ["CreateNodes - boolean schema (true)", CreateNodes, true],
  [
    "CreateArray - sub properties as boolean schema",
    CreateArray,
    {
      items: false,
      contains: false,
      prefixItems: false,
    } as JSONSchemaNS.Array,
  ],
  [
    "CreateObject - sub properties as boolean schema",
    CreateObject,
    {
      additionalProperties: false,
    } as JSONSchemaNS.Object,
  ],
  [
    "CreateObject - missing pattern in propertyNames",
    CreateObject,
    {
      propertyNames: {
        description: "I forgot the pattern key",
      },
    } as JSONSchemaNS.Object,
  ],
  [
    "CreateArray - invalid usage of Additional Items",
    CreateArray,
    {
      items: {
        type: "string",
      },
      additionalItems: {
        type: "boolean",
      },
    } as JSONSchemaNS.Array,
  ],
]

describe("JSONSchemaViewer - corner cases", () => {
  test.each(testcases)("%s", async (_title, Component, fakeSchema) => {
    // Render the component
    let rendered: RenderResult | null = null

    // Use act to ensure all updates are processed
    await act(async () => {
      rendered = render(<Component schema={fakeSchema} />)
    })

    // Capture the snapshot
    expect(rendered!.asFragment()).toMatchSnapshot()
  })

  test("Correctly infers integer when not explicitly expressed", () => {
    const foundTypes = detectedTypes({
      multipleOf: 1,
    })

    expect(foundTypes).toContain("integer")
  })
})
