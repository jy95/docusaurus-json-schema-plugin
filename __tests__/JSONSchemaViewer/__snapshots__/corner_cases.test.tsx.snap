// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - corner cases CreateArray - invalid usage of Additional Items 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    array
  </span>
  <ul>
    <li
      data-level="1"
      id="/items"
    >
      <details>
        <summary>
          <code>
            items[x]
          </code>
          Â 
          <span
            style="opacity: 0.6;"
          >
            string
          </span>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <span
              style="opacity: 0.6;"
            >
              string
            </span>
            <div />
          </div>
        </div>
      </details>
    </li>
  </ul>
  <ul>
    <li
      data-level="1"
      id="/additionalItems"
    >
      <details>
        <summary>
          <code>
            items[1,...]
          </code>
          Â 
          <span
            style="opacity: 0.6;"
          >
            boolean
          </span>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <span
              style="opacity: 0.6;"
            >
              boolean
            </span>
            <div />
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div />
</DocumentFragment>
`;

exports[`JSONSchemaViewer - corner cases CreateArray - sub properties as boolean schema 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    array
  </span>
  <ul>
    <li
      data-level="1"
      id="/prefixItems"
    >
      <details>
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span
            style="opacity: 0.6;"
          >
            none
          </span>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <span
              style="opacity: 0.6;"
            >
              none
            </span>
            <div>
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <ul>
    <li
      data-level="1"
      id="/items"
    >
      <details>
        <summary>
          <code>
            items[x]
          </code>
          Â 
          <span
            style="opacity: 0.6;"
          >
            none
          </span>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <span
              style="opacity: 0.6;"
            >
              none
            </span>
            <div>
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <ul>
    <li
      data-level="1"
      id="/contains"
    >
      <details>
        <summary>
          <code>
            items[..., x, ...]
          </code>
          Â 
          <span
            style="opacity: 0.6;"
          >
            none
          </span>
        </summary>
        <div>
          <div>
            <strong>
              type
            </strong>
            Â :Â 
            <span
              style="opacity: 0.6;"
            >
              none
            </span>
            <div>
              Always invalid
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div>
    <div>
      ðŸš¨Â No extra item(s) are authorized in this array
    </div>
  </div>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - corner cases CreateNodes - boolean schema (false) 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    none
  </span>
  <div>
    Always invalid
  </div>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - corner cases CreateNodes - boolean schema (true) 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    any
  </span>
  <div>
    Always valid
  </div>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - corner cases CreateObject - missing pattern in propertyNames 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    object
  </span>
  <div />
</DocumentFragment>
`;

exports[`JSONSchemaViewer - corner cases CreateObject - sub properties as boolean schema 1`] = `
<DocumentFragment>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style="opacity: 0.6;"
  >
    object
  </span>
  <div>
    <div>
      ðŸš¨Â No extra propertie(s) are authorized in this object
    </div>
  </div>
</DocumentFragment>
`;
